cmd/local-artifact-mirror/serve.go

```diff
@@ -25,81 +25,96 @@ var (
 
 // serveCommand starts a http server that serves files from the data directory. This server listen
 // only on localhost and is used to serve files needed by the autopilot during an upgrade.
-var serveCommand = &cli.Command{
-	Name:  "serve",
-	Usage: "Serve files from the data directory over HTTP",
-	Flags: []cli.Flag{
-		&cli.StringFlag{
-			Name:    "data-dir",
-			Usage:   "Path to the data directory",
-			Value:   ecv1beta1.DefaultDataDir,
-			EnvVars: []string{"LOCAL_ARTIFACT_MIRROR_DATA_DIR"},
-		},
-		&cli.StringFlag{
-			Name:    "port",
-			Usage:   "Port to listen on",
-			Value:   strconv.Itoa(ecv1beta1.DefaultLocalArtifactMirrorPort),
-			EnvVars: []string{"LOCAL_ARTIFACT_MIRROR_PORT"},
+func ServeCmd(ctx context.Context, v *viper.Viper) *cobra.Command {
+	var (
+		dataDir string
+		port    int
+	)
+
+	cmd := &cobra.Command{
+		Use:   "serve",
+		Short: "Serve files from the data directory over HTTP",
+		PreRunE: func(cmd *cobra.Command, args []string) error {
+			v.BindPFlag("data-dir", cmd.Flags().Lookup("data-dir"))
+			v.BindPFlag("port", cmd.Flags().Lookup("port"))
+
+			if os.Getuid() != 0 {
+				return fmt.Errorf("serve command must be run as root")
+			}
+			return nil
 		},
-	},
-	Before: func(c *cli.Context) error {
-		if os.Getuid() != 0 {
-			return fmt.Errorf("serve command must be run as root")
-		}
-		return nil
-	},
-	Action: func(c *cli.Context) error {
-		var provider *defaults.Provider
-		if c.IsSet("data-dir") {
-			provider = defaults.NewProvider(c.String("data-dir"))
-		} else {
-			var err error
-			provider, err = cmdutil.NewProviderFromFilesystem()
-			if err != nil {
-				panic(fmt.Errorf("unable to get provider from filesystem: %w", err))
+		RunE: func(cmd *cobra.Command, args []string) error {
+			v := viper.GetViper()
+
+			// there are some env vars that we still support for backwards compatibility
+			if os.Getenv("LOCAL_ARTIFACT_MIRROR_PORT") != "" {
+				envvarPort, err := strconv.ParseInt(os.Getenv("LOCAL_ARTIFACT_MIRROR_PORT"), 10, 32)
+				if err != nil {
+					return fmt.Errorf("unable to parse LOCAL_ARTIFACT_MIRROR_PORT: %w", err)
+				}
+				port = int(envvarPort)
+			}
+			if os.Getenv("LOCAL_ARTIFACT_MIRROR_DATA_DIR") != "" {
+				dataDir = os.Getenv("LOCAL_ARTIFACT_MIRROR_DATA_DIR")
 			}
-		}
 
-		port, err := k8snet.ParsePort(c.String("port"), false)
-		if err != nil {
-			panic(fmt.Errorf("unable to parse port from flag: %w", err))
-		}
+			var provider *defaults.Provider
+			if v.Get("data-dir") != nil {
+				provider = defaults.NewProvider(v.GetString("data-dir"))
```

Comment:
This does not respect the env var `LOCAL_ARTIFACT_MIRROR_DATA_DIR`. You need to use the `dataDir` variable

Reply:
i think we could really benefit from moving the flag and env var parsing to functions and write unit tests